pkg/registry/core/project/strategy.go

```diff
@@ -39,21 +40,78 @@ func (projectStrategy) NamespaceScoped() bool {
 	return false
 }
 
-func (projectStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+func (projectStrategy) PrepareForCreate(_ context.Context, obj runtime.Object) {
 	project := obj.(*core.Project)
 
 	project.Generation = 1
 	project.Status = core.ProjectStatus{}
+
+	mergeDuplicateMembers(project)
 }
 
-func (projectStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+func (projectStrategy) PrepareForUpdate(_ context.Context, obj, old runtime.Object) {
 	newProject := obj.(*core.Project)
 	oldProject := old.(*core.Project)
 	newProject.Status = oldProject.Status
 
 	if !apiequality.Semantic.DeepEqual(oldProject.Spec, newProject.Spec) {
 		newProject.Generation = oldProject.Generation + 1
 	}
+
+	mergeDuplicateMembers(newProject)
+}
+
+// TODO: This code is needed now that we have introduced validation that forbids specifying duplicates in the
+// spec.members list (this code wasn't there before). Hence, we have to remove the duplicates now to not break the API
+// incompatibly.
+// This code can be removed in a future version.
+func mergeDuplicateMembers(project *core.Project) {
+	var (
+		oldMembersToRoles = make(map[string]map[string]struct{})
+		memberToNewRoles  = make(map[string][]string)
+		newMembers        []core.ProjectMember
+	)
+
+	for _, member := range project.Spec.Members {
+		apiGroup, kind, namespace, name, err := validation.ProjectMemberProperties(member)
+		if err != nil {
+			// No meaningful way to handle the error here
+			continue
+		}
+		id := validation.ProjectMemberId(apiGroup, kind, namespace, name)
+
+		if _, ok := oldMembersToRoles[id]; !ok {
+			newMembers = append(newMembers, core.ProjectMember{
+				Subject: rbacv1.Subject{
+					APIGroup:  apiGroup,
+					Kind:      kind,
+					Namespace: namespace,
+					Name:      name,
+				},
+			})
+			oldMembersToRoles[id] = make(map[string]struct{})
+		}
+
+		for _, role := range member.Roles {
+			if _, ok := oldMembersToRoles[id][role]; !ok {
+				memberToNewRoles[id] = append(memberToNewRoles[id], role)
+			}
```

Comment:
Can you simplify the computation of the roles for the new members?
I think, it would be easier to read, if you would drop these lines and drop `memberToNewRoles` and rather loop over the keys of `oldMembersToRoles` for each member at the end.

Reply:
As discussed in person, this could potentially lead to a different order of the roles which breaks the API.